web/src/main/java/org/openmrs/web/filter/update/UpdateFilter.java

```diff
@@ -334,54 +334,116 @@ protected boolean authenticateAsSuperUser(String usernameOrSystemId, String pass
+
+			}
+			catch (Exception statementEx) {
+				if (statement != null) {
+					try {
+						statement.close();
+					}
+					catch (Exception statementCloseEx) {
+						log.error("Failed to quietly close Statement", statementCloseEx);
+					}
 				}
 			}
 			// Close statement
 			statement.close();
 		}
-		catch (Exception e) {
+		catch (Exception connectionEx) {
 			log
 			        .error(
 			            "Error while trying to authenticate as super user. Ignore this if you are upgrading from OpenMRS 1.5 to 1.6",
-			            e);
+			            connectionEx);
 			
 			// we may not have upgraded User to have retired instead of voided yet, so if the query above fails, we try
 			// again the old way
-			try {
+			if (connection != null) {
 				String select = "select user_id, password, salt from users where (username = ? or system_id = ?) and voided = '0'";
-				PreparedStatement statement = connection.prepareStatement(select);
-				statement.setString(1, usernameOrSystemId);
-				statement.setString(2, usernameOrSystemId);
-				
-				if (statement.execute()) {
-					ResultSet results = statement.getResultSet();
-					if (results.next()) {
-						Integer userId = results.getInt(1);
-						DatabaseUpdater.setAuthenticatedUserId(userId);
-						String storedPassword = results.getString(2);
-						String salt = results.getString(3);
-						String passwordToHash = password + salt;
-						return Security.hashMatches(storedPassword, passwordToHash) && isSuperUser(connection, userId);
+				PreparedStatement statement = null;
+				try {
+					statement = connection.prepareStatement(select);
+					statement.setString(1, usernameOrSystemId);
+					statement.setString(2, usernameOrSystemId);
+					if (statement.execute()) {
+						ResultSet results = null;
+						try {
+							results = statement.getResultSet();
+							if (results.next()) {
+								Integer userId = results.getInt(1);
+								DatabaseUpdater.setAuthenticatedUserId(userId);
+								String storedPassword = results.getString(2);
+								String salt = results.getString(3);
+								String passwordToHash = password + salt;
+								boolean result = Security.hashMatches(storedPassword, passwordToHash)
+								        && isSuperUser(connection, userId);
+								connection.close();
+								statement.close();
+								results.close();
+								return result;
+							} else {
+								statement.close();
+							}
+						}
+						catch (Exception resultsEx) {
+							if (results != null) {
+								try {
+									results.close();
+								}
+								catch (Exception resultsCloseEx) {
+									log.error("Failed to quietly close ResultSet", resultsCloseEx);
+								}
+							}
+						}
+					}
+				}
+				catch (Exception statementEx) {
+					if (statement != null) {
+						try {
+							statement.close();
+						}
+						catch (Exception statementCloseEx) {
+							log.error("Failed to quietly close Statement", statementCloseEx);
+						}
 					}
 				}
-			}
-			catch (Exception t2) {
```

Comment:
It's replaced with the if statement above, I think it was there to check on the
connection that actually threw the exception.